staging/src/k8s.io/sample-controller/pkg/signals/signal.go

```diff
@@ -26,18 +27,18 @@ var onlyOneSignalHandler = make(chan struct{})
 // SetupSignalHandler registered for SIGTERM and SIGINT. A stop channel is returned
 // which is closed on one of these signals. If a second signal is caught, the program
 // is terminated with exit code 1.
-func SetupSignalHandler() (stopCh <-chan struct{}) {
+func SetupSignalHandler() (context.Context, context.CancelFunc) {
```

Comment:
The caller doesn't need the cancel function. That will be called internally by the goroutine.

Reply:
Can you elaborate on your statement "internally by the goroutine" ? Which goroutine ?

There seems to be no goroutine waiting for <-ctx.Done() except the main goroutine (in Run function). So by the time we call the deferred close in main we are guaranteed that no goroutine will be waiting on ctx.Done. So the cancel function is not needed. Or is there another reason to remove this ?

Reply:
The one started by `go func() {` inside this function. That's the one which calls `cancel`. The caller then just consumes the `ctx`.

Reply:
The cancel is called only when SIGINT or SIGTERM is received, there is an alternate path through the main function which will never hit this cancel (if no SIGINT/SIGTERM is received). But here is no other instances of goroutines waiting on ctx->Done in that path.

Reply:
Fixed in the latest commit.